cbmc/proofs/IotMqtt_UnsubscribeSync/IotMqtt_UnsubscribeSync_harness.c

```diff
@@ -0,0 +1,161 @@
+/*
+ * IoT MQTT V2.1.0
+ * Copyright (C) Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person
+ * obtaining a copy of this software and associated documentation
+ * files (the "Software"), to deal in the Software without
+ * restriction, including without limitation the rights to use, copy,
+ * modify, merge, publish, distribute, sublicense, and/or sell copies
+ * of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be
+ * included in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+ * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+ * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+ * NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS
+ * BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
+ * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+/**
+ * @file IotMqtt_UnsubscribeSync_harness.c
+ * @brief Implements the proof harness for IotMqtt_UnsubscribeSync function.
+ */
+
+#include "iot_config.h"
+#include "private/iot_mqtt_internal.h"
+
+#include <stdlib.h>
+
+#include "mqtt_state.h"
+
+/****************************************************************/
+
+/**
+ * We abstract this function for performance reasons. In its original
+ * implementation, CBMC ended up creating byte extracts for all possible
+ * objects, due to the polymorphic nature of the linked list. We assume
+ * that the function is memory safe, and we free and havoc the list to
+ * demonstrate that no subsequent code makes use of the values in the list.
+ */
+
+typedef bool ( *MatchFunction_t )( const IotLink_t * const pOperationLink,
+                                   void * pCompare );
+typedef void ( *FreeElementFunction_t )( void * pData );
+
+void IotListDouble_RemoveAllMatches( const IotListDouble_t * const pList,
```

Comment:
Is this common to the past few proofs I've reviewed? Shouldn't it abstracted to some shared module then?

Reply:
I'd like to get all proofs done and in the repository (4 of 22 remain after this pull request).  Then I'd like to do an audit of all the proofs, factor out common code, and make all styles and formats consistent.  Right now, though, I'd just like to get the proofs in.